File: src/main/java/systems/conduit/main/core/plugin/config/ConfigurationTypes.java
package systems.conduit.main.core.plugin.config;

import systems.conduit.main.Conduit;
import systems.conduit.main.core.plugin.config.loader.JsonLoader;

import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

/*
 * @author Innectic
 * @since 10/17/2019
 */
public class ConfigurationTypes {
    private static final Map<String, ConfigurationLoader> loaderCache = new HashMap<>();
    private static final Map<String, Class<? extends ConfigurationLoader>> loaderTypes = new HashMap<>();

    static {
        loaderTypes.put("json", JsonLoader.class);
    }

    public static Optional<ConfigurationLoader> getLoaderForExtension(String extension) {
        // First, check if we currently have an already created loader cached.
        if (loaderCache.containsKey(extension)) return Optional.of(loaderCache.get(extension));

        // Since it does not currently contain an instance, lets attempt to make one.
        Class<? extends ConfigurationLoader> loaderClass = loaderTypes.getOrDefault(extension, null);
        // Invalid extension provided.
        if (loaderClass == null) return Optional.empty();

        // Valid extension, so we have a class type and can make a new cache loader.
        try {
            ConfigurationLoader loader = loaderClass.getConstructor().newInstance();
            loaderCache.put(extension, loader);
            return Optional.of(loader);
        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
            Conduit.getLogger().error("Failed to create instance of configuration loader for " + extension + ":");
            e.printStackTrace();
        }
        return Optional.empty();
    }

    public static void drop(String extension) {
        loaderCache.remove(extension);
    }
}

File: src/main/kotlin/systems/conduit/core/plugin/config/ConfigurationTypes.kt
package systems.conduit.core.plugin.config

import systems.conduit.core.Conduit
import systems.conduit.core.plugin.config.loader.JsonLoader
import java.lang.reflect.InvocationTargetException
import java.util.*

/*
 * @author Innectic
 * @since 10/17/2019
 */
object ConfigurationTypes {

    private val loaderCache: MutableMap<String, ConfigurationLoader> = HashMap()
    private val loaderTypes: MutableMap<String, Class<out ConfigurationLoader>?> = HashMap()

    fun getLoaderForExtension(extension: String): ConfigurationLoader? {
        // First, check if we currently have an already created loader cached.
        if (loaderCache.containsKey(extension)) return loaderCache[extension]!!

        // Since it does not currently contain an instance, lets attempt to make one.
        val loaderClass = loaderTypes.getOrDefault(extension, null) ?: return null
        // Invalid extension provided.

        // Valid extension, so we have a class type and can make a new cache loader.
        try {
            val loader = loaderClass.getConstructor().newInstance()
            loaderCache[extension] = loader
            return loader
        } catch (e: InstantiationException) {
            Conduit.logger.error("Failed to create instance of configuration loader for $extension:")
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            Conduit.logger.error("Failed to create instance of configuration loader for $extension:")
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            Conduit.logger.error("Failed to create instance of configuration loader for $extension:")
            e.printStackTrace()
        } catch (e: NoSuchMethodException) {
            Conduit.logger.error("Failed to create instance of configuration loader for $extension:")
            e.printStackTrace()
        }
        return null
    }

    fun drop(extension: String) {
        loaderCache.remove(extension)
    }

    init {
        loaderTypes["json"] = JsonLoader::class.java
    }
}